cpp/src/io/utilities/column_buffer.hpp

```diff
@@ -108,25 +85,111 @@ struct column_buffer {
 
   auto& null_count() { return _null_count; }
 
-  // Create a new column_buffer that has empty data but with the same basic information as the
-  // input column, including same type, nullability, name, and user_data.
-  static column_buffer empty_like(column_buffer const& input);
-
-  std::unique_ptr<rmm::device_uvector<string_index_pair>> _strings;
   rmm::device_buffer _data{};
   rmm::device_buffer _null_mask{};
   size_type _null_count{0};
 
   data_type type{type_id::EMPTY};
   bool is_nullable{false};
   size_type size{0};
-  std::vector<column_buffer> children;
   uint32_t user_data{0};  // arbitrary user data
   std::string name;
 
   rmm::mr::device_memory_resource* mr;
 };
 
+struct column_buffer_with_pointers : public column_buffer_base {
+  column_buffer_with_pointers() = default;
+
+  // construct without a known size. call create() later to actually allocate memory
+  column_buffer_with_pointers(data_type _type, bool _is_nullable)
+    : column_buffer_base(_type, _is_nullable)
+  {
+  }
+
+  void create(size_type _size, rmm::cuda_stream_view stream, rmm::mr::device_memory_resource* _mr);
+
+  void* data() { return _strings ? _strings->data() : _data.data(); }
+  size_t data_size() const { return _strings ? _strings->size() : _data.size(); }
+
+  void create_string_data(size_t num_bytes, rmm::cuda_stream_view stream)
+  {
+    CUDF_FAIL("method not implemented for type");
+  }
+
+  void* string_data() { CUDF_FAIL("method not implemented for type"); }
+  size_t string_size() const { CUDF_FAIL("method not implemented for type"); }
+
+  std::unique_ptr<column> make_string_column(rmm::cuda_stream_view stream);
+
+  std::unique_ptr<rmm::device_uvector<string_index_pair>> _strings;
+};
+
+struct column_buffer_with_strings : public column_buffer_base {
+  column_buffer_with_strings() = default;
+
+  // construct without a known size. call create() later to actually allocate memory
+  column_buffer_with_strings(data_type _type, bool _is_nullable)
+    : column_buffer_base(_type, _is_nullable)
+  {
+  }
+
+  void create(size_type _size, rmm::cuda_stream_view stream, rmm::mr::device_memory_resource* _mr);
+
+  void* data() { return _data.data(); }
+  size_t data_size() const { return _data.size(); }
+
+  void create_string_data(size_t num_bytes, rmm::cuda_stream_view stream);
+
+  void* string_data() { return _string_data.data(); }
+  size_t string_size() const { return _string_data.size(); }
+
+  std::unique_ptr<column> make_string_column(rmm::cuda_stream_view stream);
+
+  rmm::device_buffer _string_data{};
+};
+
+/**
+ * @brief Class for holding device memory buffers to column data that eventually
+ * will be used to create a column.
+ */
+template <class string_policy>
+struct column_buffer : string_policy {
```

Comment:
I think we can remove the need to separate column_buffer and column_buffer_base classes with CRTP. If I understand the purpose of current column_buffer correctly, this should work:
```
template <typename string_policy>
class column_buffer {
// implement non-string stuff

// I think SFINAE can be used here to compile-time disable some members
void string_data() {
        static_cast<T*>(this)->string_data();
}
...

std::vector<column_buffer<string_policy>> children;
};

class column_buffer_with_pointers : column_buffer<column_buffer_with_pointers>

Reply:
Curious how you keep repeating your desire for CRTP :rofl: If you don't get to it this week I'll give it a try in a bit (might get pulled onto other work today).

Reply:
I'll try to make a commit locally today.